### PR TITLE
[QMS-694] Porting Qt6: Fix decoding of device mount point

### PR DESCRIPTION
**Pull request by kiozen**
### What is the linked issue for this pull request:
[comment]: # (Add the ticket number behind QMS-# )

QMS-#694

### What you have done:
[comment]: # (Describe roughly.)

Refactured code to decode DBus messages properly

### Steps to perform a simple smoke test:
[comment]: # (List the steps.)

Check if a Garmin device can be mounted and read

### Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):

- [x] yes

### Is every user facing string in a tr() macro?

- [x] yes

### Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.

- [ ] yes, I didn't forget to change changelog.txt

### CODE REVIEW

File: src/qmapshack/device/CDeviceWatcherLinux.cpp [hunk 1]
```diff
@@ -157,30 +156,54 @@ QString CDeviceWatcherLinux::readMountPoint(const QString& path) {
...
+    dbusArg.beginArray();
+    while (!dbusArg.atEnd()) {
+      QByteArray data;
+      dbusArg >> data;
+      // Not sure about this one. But my guts feeling tells me that
+      // we should decode the stuff according to the system's requirements.
+      auto decoder = QStringDecoder(QStringDecoder::System);
+      // As QByteArray has a trailing 0 however QString does not it needs to be
+      // removed. Else the QString has a 0 as unicode character, confusing the
+      // rest of the system.
```
Comment (frankystone) [inline src/qmapshack/device/CDeviceWatcherLinux.cpp:198]: This sounds like the `0` is added by QByteArray, but it's the last entry in the array coming from the DBus.

**Review by frankystone — COMMENTED** [review]
Works!

Thanks for spending your time on this.

File: src/qmapshack/device/CDeviceWatcherLinux.cpp (hunk 1 above)
Comment (kiozen) [reply src/qmapshack/device/CDeviceWatcherLinux.cpp:198]: According to the docs QByteArray always enforces a trailing 0 . If there is one in the raw data (DBus message) it's ok. If not it is appended by QByteArray. So it'S always guaranteed to be 0 terminated. 

The part I do not understand is why QString does not recognize it as 0 terminated string and takes over the 0 as part of the string.

Comment (frankystone) [reply src/qmapshack/device/CDeviceWatcherLinux.cpp:198]: From my understanding QByteArray represents just a number of pure bytes, whereas QString is always UTF16 encoded. From [Using QByteArray](https://wiki.qt.io/Using_QByteArray):

> To convert a QByteArray to QString, one needs to specify the encoding. It is not possible to convert a QByteArray to QString without knowing the encoding. Once the encoding is known, one can use functions like QString::fromUtf8, QString::fromUcs4 or use QTextCodec::toUnicode().

But i have no experience in QT.

Comment (kiozen) [reply src/qmapshack/device/CDeviceWatcherLinux.cpp:198]: Yes decoding is another caveat when converting from QByteArray to QString. That is why I introduced the decoder. Anyways this does not explain why the trailing 0 is converted to an explicit "\0" as part of the string.

Comment (frankystone) [reply src/qmapshack/device/CDeviceWatcherLinux.cpp:198]: You should ask this in a QT forum…

In python it's the same, i used the int's from my qdbus output:
~~~~
>>> message = [47, 114, 117, 110, 47, 109, 101, 100, 105, 97, 47, 107, 97, 112, 117, 116, 116, 110, 105, 107, 47, 75, 65, 82, 84, 69, 32, 54, 48, 71, 66, 0]
>>> ar = bytearray(message)
>>> print(ar)
bytearray(b'/run/media/kaputtnik/KARTE 60GB\x00')
>>> s = str(ar)
>>> print(s)
bytearray(b'/run/media/kaputtnik/KARTE 60GB\x00')
>>> print(ar.decode())
/run/media/kaputtnik/KARTE 60GB
>>> s = str(ar, encoding='utf-8')
>>> print(s)
/run/media/kaputtnik/KARTE 60GB
~~~~

Comment (kiozen) [reply src/qmapshack/device/CDeviceWatcherLinux.cpp]: I think the correct way to access the byte array is to use the `const char*` pointer. By that decoding can detect the null termination properly.

**Comment by kiozen** [comment]
@wthaem does this work for Windows, too?

**Comment by kkarsten62** [comment]
I compiled kiozen qmapshack fork with branch [QMS-694] against `| Fedora 41 | Wayland | Gnome 47 | Qt 6.8.0`.

**Work as expected.**

I can read/write from/to my GARMIN Edge 1000 device.

**Comment by wthaem** [comment]
@kiozen: I'll test it in the next few days.

**Comment by kiozen** [comment]
@wthaem I think it's enough to test the porting branch on Windows in general. Not necessarily this branch. I just want to make sure we do not have similar issues in the Windows code.

**Comment by wthaem** [comment]
@kiozen: I mentioned in #648 a failing string compare issue in `CMapRMAP.cpp`. There, an extra zero byte played a role. I think that similar zero byte, QString, QArray, or encoding issues can be hidden elsewhere in the code, too.

**Comment by kiozen** [comment]
> @kiozen: I mentioned in #648 a failing string compare issue in `CMapRMAP.cpp`. There, an extra zero byte played a role. I think that similar zero byte, QString, QArray, or encoding issues can be hidden elsewhere in the code, too.

Ok, can you check if `"CompeGPSRasterImage" != QString(charbuf.data())` does the trick.

**Comment by wthaem** [comment]
Tested QMS with Qt6 on Windows 10:

@kiozen:

* `CMapRMAP.cpp` - `"CompeGPSRasterImage" != QString(charbuf.data())`:  Could load rmap maps. Seems to work. How to identify similar situations elsewhere in the code?
* I connected my Garmin GPSMap 64s to QMS. Works in read and write modes without any problems.
* Currently, I use the QMS Qt6 version for my normal work with QMS. Thus, more or less frequent testing of the features which are used in my workflow.

**Comment by kiozen** [comment]
I merge this one and open a new one to collect all other QByteArray fixes

**Comment by kiozen** [comment]
That is the new ticket https://github.com/Maproom/qmapshack/issues/697

**Pull request merged**